Tidying up (No functional changes)

Senryoku opened the pull request:
A couple of things I noticed while browsing the code (and before realizing the changes I wanted to propose were actually already implemented :^) )

- Convert some `reinterpret_cast` into more appropriate `dynamic_cast`s for downcasting in a virtual hierarchy.
- Avoid any string allocation in FractionalNumber until they're actually used. 
  I noticed this type is used for arithmetic operations in the rate update and doing a bunch of unnecessary allocation there, at least in debug mode.

Sorry for the boring PR and thanks for making and publishing this tool, I'll definitely use it when 1.0 finally comes out!

Senryoku committed:
Remove some unnecesary reinterpret_cast

Senryoku committed:
FractionalNumber: Avoid string allocation until necessary by wrapping them in optionals.

This allows using this type for arithmetic purposes without paying the cost of string allocations.

adepierre commented:
Hello, thank you for the PR!

I think reinterprete_cast is a better fit in this case, we already use polymorphism to check the type of the node using the virtual GetKind() function. So for me there is no real reason here to use the slower dynamic_cast (not that it makes a real performance difference overall but on principle).

Regarding the std::optional for the strings, I'm curious what's the reason behind it? You save the std::string 32 bytes but instead you create a std::optional that takes 40 bytes.

Senryoku commented:
> Hello, thank you for the PR!

Hi, my pleasure! :)

> I think reinterprete_cast is a better fit in this case, we already use polymorphism to check the type of the node using the virtual GetKind() function. So for me there is no real reason here to use the slower dynamic_cast (not that it makes a real performance difference overall but on principle).

I'd only use `reinterpret_cast` when absolutely necessary because it doesn't show a clear intent (and the intended use cases can probably even be replaced by a `std::bit_cast` nowadays I guess?). I choose `dynamic_cast` here because I think it's technically the safer option, but you're right, it's overkill if you can guarantee that the type is correct. TBH, I would have used `static_cast` directly here if this was my own project and didn't want to be extra cautious.

> Regarding the std::optional for the strings, I'm curious what's the reason behind it? You save the std::string 32 bytes but instead you create a std::optional that takes 40 bytes.

Yes, it's larger, but it prevents `std::string` initialization both at construction and every time the value changes (the assignment in `FractionalNumber::UpdateValue`). I assumed small string optimization would prevent this but I was seeing a lot of allocations and deallocation in `UpdateNodesRate` - maybe it is not used in debug build?
If it is actually beneficial will depend on the `std::string` implementation, but I also think this makes the intent clearer, and it guarantees this won't be a problem if the STL implementation happens to allocate unconditionally when constructing a `std::string`.

Senryoku committed:
Convert dynamic_cast to static_cast.

Types are known, no need to add runtime checks. At least we get comptime checks compared to reinterpret_cast :)

Senryoku commented:
I changed the casts again to `static_cast`s (including one that was previously a `dynamic_cast`, while properly guarded by the value of `GetKind()`).
No runtime difference compared to `reinterpret_cast`, but at least guarantees that the types are compatible at compile time (who knows, maybe it'll help someday when refactoring).

adepierre commented:
Yeah looks fine to me thanks.

For the strings I'm not really worried about the performances for the desktop version as it is currently running on probably even a potato. I'm a bit more concerned about the web version as I have literally no idea how to profile this after wasm compilation.

Anyway, I agree with the intent being clearer with the optional.
